Specify active service worker inheritance

jungkees opened the pull request:
This adds steps to make clients that have creator browsing contexts to
inherit the active service worker from their creators. While they have
an initial inherited active service worker, if one exists, navigation
matching still always wins. So, if the navigation goes through HTTP
fetch, the inherited active service worker is replaced by the matched
result. Also, if an iframe has a sandbox attribute set without
allow-same-origin token, the inherited active service worker is set to
null.

Related issue: https://github.com/w3c/ServiceWorker/issues/765.

domenic commented:
This LGTM editorially, and I will trust you on the semantics. Is someone slated to write web platform tests for this?

jakearchibald commented on source:
```diff
@@ -77882,6 +77887,16 @@ console.assert(iframeWindow.frameElement === null);
+   context</span>, then set <var>settings object</var>'s <span
+   data-x="concept-environment-active-service-worker">active service worker</span> to <span>creator
+   browsing context</span>'s <span>active document</span>'s <span>relevant settings object</span>'s
+   <span data-x="concept-environment-active-service-worker">active service worker</span>.</p></li>
```
If I have an iframe pointing to another origin, am I right in reading it'll initially inherit my service worker via this step, but will discard it during navigation? If the target origin doesn't have a service worker, when is the active worker set to null?

jakearchibald commented:
Do we need similar steps for "set up a worker environment settings object" or do we handle inheritance for workers elsewhere?

jungkees commented on source:
```diff
@@ -77882,6 +77887,16 @@ console.assert(iframeWindow.frameElement === null);
    </li>
 
+   <li><p>Set <var>settings object</var>'s <span
+   data-x="concept-environment-active-service-worker">active service worker</span> to null.</p></li>
```
For the clients that don't get to http fetch, the following steps set the value:

> set <var>settings object</var>'s <span data-x="concept-environment-active-service-worker">active service worker</span> to <span>creator browsing context</span>'s <span>active document</span>'s <span>relevant settings object</span>'s <span data-x="concept-environment-active-service-worker">active service worker</span>.

and for the clients that get through http fetch, the following steps set the value:

> , and <var>settings object</var>'s <span data-x="concept-environment-active-service-worker">active service worker</span> to <var>reserved environment</var>'s <span
     data-x="concept-environment-active-service-worker">active service worker</span>.

the reserved client's active service worker here is set in Handle Fetch where the registration matching is done.

jungkees commented:
I think I missed the worker case. When a script URL for a worker is a local URL, it doesn't get to http fetch. So, currently that worker client won't get a controller. (i.e. it's just null.) We have related text here - https://w3c.github.io/ServiceWorker/#selection - but I think we should add relevant steps to "set up a worker environment settings object". It should inherit either worker's document's controller or worker's worker's controller in this case I suppose.

domenic commented:
@jungkees @jakearchibald what is the latest on this? Is anyone writing tests for it---perhaps there are tests in Blink we can import?

annevk commented:
> That later-set "about:blank" document can even have a different origin to its parent.

Really? Do you have a test for that? Would be quite surprising.

jungkees commented:
> Really? Do you have a test for that? Would be quite surprising.

It was my misunderstanding. The [definition of document's origin](https://html.spec.whatwg.org/#concept-origin) says "If the Document is a non-initial "about:blank" document," The origin is determined as  "the **incumbent** settings object when the navigate algorithm was invoked..". I thought the incumbent settings object can be different from the top-level script's environment settings object, but having checked the incumbent concept it seems the incumbent settings object is always the parent's environment setting object in this case.

Testing with the following code in Chrome says they're the same. (Firefox doesn't seem to offer document.origin yet.):
```js
new Promise(resolve => {
  let frame = document.createElement('iframe');
  frame.src = 'blank.html';
  frame.onload = _ => { resolve(frame); };
  document.body.appendChild(frame);
}).then(frame => {
  return new Promise(resolve => {
     console.log(frame.contentWindow.document.origin);
     frame.src = 'about:blank';
     frame.onload = _ => { resolve(frame); };
  });
}).then(frame => {
  console.log(frame.contentWindow.document.origin);
});
```

So, should we inherit the parent's controller in this case? The patches in this PR currently set the controller to null as it goes through [process a navigate fetch](https://html.spec.whatwg.org/#process-a-navigate-fetch) - meaning it creates a reserved client - and returned with no matched SW.

jungkees commented:
I found another corner case that needs clarification. With the following scenario:

1. A document whose origin is foo.com is loaded in the parent frame.
1. An iframe with no src set is inserted in the parent document. // Initial about:blank document's origin is inherited from the parent: foo.com.
1. Navigate the iframe's nested browsing context to foo.com/resource.

When the navigate reaches [Initialzing a new Document object](https://html.spec.whatwg.org/#initialise-the-document-object) step 3, the new Document reuses the Window object and the environment settings object of the initial Document rather than creating its own ones. That is, the new Document will use the active SW inherited from the parent document even if it has gotten a matched SW through the http fetch.

For this case, I think we should replace the inherited active SW with the matched SW by which the main resource has already been served. Does it makes sense? And it seems other environment's setup (e.g. id, creation URL, etc.) also needs to be overriden I guess.

Is there any particular reason the existing algorithm reuses the Window and the environment settings object only in this case in the first place?

annevk commented:
I think that is to avoid losing global variables and getting too many history entries. @bzbarsky and maybe @smaug---- could say more about it. We really need to refactor how all these things get created and teared down, since it's a giant mess.

smaug---- commented:
losing global variables and losing event listeners would be bad.

bzbarsky commented:
> Is there any particular reason the existing algorithm reuses the Window

Historically, because people have set variables and event listeners on the window before it's done loading, racing with the load getting its initial HTTP response headers (and the document being created).  And browsers made that work regardless of whether the race was won or lost.

It's very much worth investigating what actual browser behavior around this stuff is.  The spec was written more or less to match Gecko, but I know for a fact Chrome/WebKit have different (much more confusing, which is why it was not specced) behavior, and Edge/IE have yet a third behavior in the window.open case; dunno about the iframe case.

It's _possible_ that there are web-compatible simplifications to the behavior here that would make all this a non-issue, though unlikely: Hixie is pretty good at testing various edge-cases...

jungkees commented:
@smaug----, @bzbarsky, thanks for the answers. 
> It's very much worth investigating what actual browser behavior around this stuff is.

It surely is an area that needs investigation.

@domenic, @jakearchibald, @annevk,

I pushed a new commit to cover two corner cases that were mentioned in the above comments:
 - For navigating an iframe with the initial about:blank Document, I added steps to replace the inherited active SW with the matched SW (which the main resource has been served by.)
 - For navigating an iframe to a non-initial about:blank Document, I added steps to make it inherit the parent's SW rather than leaving it to null set during the navigate fetch.

And added the wpt tests: https://github.com/w3c/web-platform-tests/pull/4610.

domenic commented:
Ping reviewers @jakearchibald and @annevk; I don't feel qualified to review this beyond just editorial matters. I see @mkruisselbrink is reviewing the tests; maybe he can help as well.

jungkees commented:
I think it's difficult to rebase, review, and discuss on this PR after a long pause of work. I created a new PR for this effort. Please come to https://github.com/whatwg/html/pull/2809.
